### PR TITLE
GPT inference long context

### PR DESCRIPTION
**Pull request by ekmb**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py [hunk 1]
```diff
@@ -169,7 +169,7 @@ def _process_example(self, example):
         tokenized_text = pre_pad + self.tokenizer.text_to_ids(text)
         context_ids = pre_pad + self.tokenizer.text_to_ids(context)
         answer_ids = tokenized_text[len(context_ids) :]
-        total_ids = len(context_ids) + len(answer_ids)
+        total_ids = len(context_ids) + max(len(answer_ids), self.tokens_to_generate)
```
Comment (arendu) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:172]: Why is tokens to generate added here?  Does this not suffice? https://github.com/NVIDIA/NeMo/blob/20717bab25f5d0ebe4cbe7fbef614206aaa630e5/nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py#L265

File: examples/nlp/language_modeling/tuning/megatron_gpt_peft_eval.py [hunk 2]
```diff
@@ -119,6 +119,7 @@ def main(cfg) -> None:
         peft_model_cfg.data.test_ds = cfg.model.data.test_ds
         peft_model_cfg.activations_checkpoint_granularity = None
         peft_model_cfg.activations_checkpoint_method = None
+        peft_model_cfg.encoder_seq_length = cfg.model.data.test_ds.max_seq_length
```
Comment (arendu) [inline examples/nlp/language_modeling/tuning/megatron_gpt_peft_eval.py:122]: `encoder_seq_length` is a property of the model, this should ideally be updated during training if the model does support longer contexts.

File: nemo/collections/nlp/modules/common/text_generation_utils.py [hunk 3]
```diff
@@ -662,10 +662,10 @@ def sample_sequence_batch(
             output = inference_strategy.forward_step(batch, tensor_shape)
 
             if parallel_state.is_pipeline_last_stage():
-                output = output[0]['logits'].float()
+                output = output[0]['logits']  # .float()
```
Comment (arendu) [inline nemo/collections/nlp/modules/common/text_generation_utils.py:665]: @yidong72 implemented this change. Yi is it ok to not convert the outputs to fp32?

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py [hunk 4]
```diff
@@ -207,7 +207,7 @@ def _process_example(self, example):
 
         if len(input_ids) < self.min_seq_length or len(input_ids) > self.max_seq_length:
             input_ids = input_ids[: self.max_seq_length]
-
+        print("len context_ids", len(context_ids))
```
Comment (yidong72) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:210]: remove this debugging line.

File: nemo/collections/nlp/modules/common/megatron/module.py [hunk 5]
```diff
@@ -290,8 +290,8 @@
         if getattr(self.module, 'pre_process', True):
             inputs = fp32_to_float16(inputs, self.float16_converter)
         outputs = self.module(*inputs, **kwargs)
-        if parallel_state.is_pipeline_last_stage():
-            outputs = float16_to_fp32(outputs)
+        # if parallel_state.is_pipeline_last_stage():
+        #     outputs = float16_to_fp32(outputs)
```
[the remark this answers is not in the document]
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/module.py:294]: @ericharper, any reason we need the output to be in float32 type? Is it fine to comment this out?

File: nemo/collections/nlp/modules/common/text_generation_utils.py (hunk 3 above)
Comment (yidong72) [reply nemo/collections/nlp/modules/common/text_generation_utils.py:665]: For inference, I think it is fine not to use fp32.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py (hunk 1 above)
Comment (ekmb) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:174]: resolved offline

**Review by arendu — APPROVED** [review]
LGTM!

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**